File: src/main/java/kroppeb/server/command/commands/TagCommand.java
/*
 * Copyright (c) 2020 Kroppeb
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package kroppeb.server.command.commands;

import com.google.common.collect.Sets;
import kroppeb.server.command.Command;
import kroppeb.server.command.arguments.Selector;
import kroppeb.server.command.reader.Reader;
import kroppeb.server.command.reader.ReaderException;
import net.minecraft.entity.Entity;
import net.minecraft.server.command.ServerCommandSource;

import java.util.HashSet;
import java.util.Set;

abstract public class TagCommand implements Command {
	static public TagCommand read(Reader reader) throws ReaderException {
		Selector targets = Selector.read(reader);
		reader.moveNext();
		String sub = reader.readLiteral();
		switch(sub){
			case "add": return new Add(targets, reader.readUnquotedString());
			case "list": return new List(targets);
			case "remove": return new Remove(targets, reader.readUnquotedString());
			default: throw new ReaderException("Unknown tag subcommand: " + sub);
		}
	}
	
	static public class Add extends TagCommand{
		final Selector targets;
		final String name;
		
		public Add(Selector targets, String name) {
			this.targets = targets;
			this.name = name;
		}
		
		@Override
		public int execute(ServerCommandSource source) {
			int count = 0;
			for (Entity entity : targets.getEntities(source)) {
				if(entity.addScoreboardTag(name))
					count++;
			}
			return count; // TODO throw if 0?
		}
	}
	
	static public class List extends TagCommand{
		final Selector targets;
		
		public List(Selector targets) {
			this.targets = targets;
		}
		
		@Override
		public int execute(ServerCommandSource source) {
			Set<Object> tags = Sets.newHashSet();
			for (Entity entity : targets.getEntities(source)) {
				tags.addAll(entity.getScoreboardTags());
			}
			return tags.size();
		}
	}
	
	static public class Remove extends TagCommand{
		final Selector targets;
		final String name;
		
		public Remove(Selector targets, String name) {
			this.targets = targets;
			this.name = name;
		}
		
		@Override
		public int execute(ServerCommandSource source) {
			int count = 0;
			for (Entity entity : targets.getEntities(source)) {
				if(entity.removeScoreboardTag(name))
					count++;
			}
			return count; // TODO throw if 0?
		}
	}
}

File: src/main/kotlin/kroppeb/server/command/commands/TagCommand.kt
/*
 * Copyright (c) 2020 Kroppeb
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package kroppeb.server.command.commands

import com.google.common.collect.Sets
import kroppeb.server.command.Command
import kroppeb.server.command.InvocationError
import kroppeb.server.command.reader.Reader
import kroppeb.server.command.reader.ReaderException
import kroppeb.server.command.arguments.selector.Selector
import kroppeb.server.command.arguments.selector.Selector.Companion.parse
import kroppeb.server.command.reader.Literal
import kroppeb.server.command.reader.ReadFactory
import net.minecraft.server.command.ServerCommandSource

abstract class TagCommand : Command {
	class Add(val targets: Selector, val name: String?) : TagCommand() {
		override fun execute(source: ServerCommandSource): Int {
			var count = 0
			val entities = targets.getEntities(source)
			if (entities!!.isEmpty()) throw InvocationError()
			for (entity in entities) {
				if (entity!!.addScoreboardTag(name)) count++
			}
			return count
		}

	}

	class List(val targets: Selector) : TagCommand() {
		override fun execute(source: ServerCommandSource): Int {
			val entities = targets.getEntities(source)
			if (entities!!.isEmpty()) throw InvocationError()
			val tags: MutableSet<Any> = Sets.newHashSet()
			for (entity in entities) {
				tags.addAll(entity!!.scoreboardTags)
			}
			return tags.size
		}

	}

	class Remove(val targets: Selector, val name: String?) : TagCommand() {
		override fun execute(source: ServerCommandSource): Int {
			val entities = targets.getEntities(source)
			if (entities!!.isEmpty()) throw InvocationError()
			var count = 0
			for (entity in entities) {
				if (entity!!.removeScoreboardTag(name)) count++
			}
			return count
		}

	}

	companion object:ReadFactory<TagCommand> {
		override fun Reader.parse(): TagCommand {
			val targets = Selector()
			return when (val sub = Literal()) {
				"add" -> Add(targets, Literal())
				"list" -> List(targets)
				"remove" -> Remove(targets, Literal())
				else -> throw ReaderException("Unknown tag subcommand: $sub")
			}
		}
	}
}
